plugins/out_s3/s3.c

```diff
@@ -2102,35 +1940,125 @@ static void flush_init(void *out_context)
-    if (ctx->timer_created == FLB_FALSE) {
-        flb_plg_debug(ctx->ins,
-                      "Creating upload timer with frequency %ds",
-                      ctx->timer_ms / 1000);
+    set = FLB_TASK_SET(ret, task->id, o_ins->id);
+    val = FLB_BITS_U64_SET(2 /* FLB_ENGINE_TASK */, set);
 
-        sched = flb_sched_ctx_get();
+    /*
+     * Set the target pipe channel: if this return code is running inside a
+     * thread pool worker, use the specific worker pipe/event loop to handle
+     * the return status, otherwise use the channel connected to the parent
+     * event loop.
+     */
+    if (flb_output_is_threaded(o_ins) == FLB_TRUE) {
+        /* Retrieve the thread instance and prepare pipe channel */
+        th_ins = flb_output_thread_instance_get();
+        pipe_fd = th_ins->ch_thread_events[1];
+    }
+    else {
+        pipe_fd = out_flush->o_ins->ch_events[1];
+    }
 
-        if (ctx->preserve_data_ordering) {
-            ret = flb_sched_timer_cb_create(sched, FLB_SCHED_TIMER_CB_PERM,
-                                            ctx->timer_ms, s3_upload_queue, ctx, NULL);
-        }
-        else {
-            ret = flb_sched_timer_cb_create(sched, FLB_SCHED_TIMER_CB_PERM,
-                                            ctx->timer_ms, cb_s3_upload, ctx, NULL);
-        }
-        if (ret == -1) {
-            flb_plg_error(ctx->ins, "Failed to create upload timer");
-            FLB_OUTPUT_RETURN(FLB_RETRY);
+    /* Notify the event loop about our return status */
+    n = flb_pipe_w(pipe_fd, (void *) &val, sizeof(val));
+    if (n == -1) {
+        flb_errno();
+    }
+}
+
+/* 
+ * We need to use async IO for S3 because it is more stable
+ * However, S3 has unique needs. The chunk list, multipart code, etc
+ * all is not concurrent safe. 
+ * Additionally, timer callbacks don't run in coroutines and
+ * can't using async IO. 
+ * Solution: daemon coroutine
+ * The first coroutine that that flushes to S3 never ends,
+ * and just uploads and sleeps. 
+ * 
+ * We increment the metrics counters for the chunk originally
+ * associated with the coroutine and decrement the task users
+ * in release_chunk_upstream()
+ */
+static void daemon_coroutine(struct flb_config *config, struct flb_s3 *ctx)
+{
+    flb_plg_info(ctx->ins, "daemon coroutine starting...");
+
+    ctx->daemon_coro_started = FLB_TRUE;
+
+    /* tell engine that this task did complete successfully */
+    flb_output_return_no_destroy(FLB_OK);
+
+    /* 
+     * FLB engine uses a graceful cooperative shutdown model. 
+     * If coroutines never end, the system won't stop.
+     * So the daemon coroutine must exit itself when the engine is in shutdown mode.
+     */
+    while (config->is_running == FLB_TRUE) {
+        /* Cleanup old buffers found on startup */
+        flush_startup_chunks(ctx);
+
+        /* upload any ready chunks */
+        cb_s3_upload(config, ctx);
+
+        if (config->is_running == FLB_FALSE) {
+            break;
         }
```

Comment:
This looks good. Do you know if retries are not processed after the grace timeout has expired?

I see ->is_running is set in the flb_engine_shutdown function after grace timeout is triggered (as mentioned in the comment)
https://github.com/fluent/fluent-bit/blob/1.9/src/flb_engine.c

If retries can still be triggered, however, then we may have an edge case where a new chunk could be ingested.

Reply:
## Concern 2
If there are failures in cb_s3_upload, then we may want this daemon thread to continue running. In that case, the while loop condition could be:

```
while (config->is_running == FLB_TRUE || <there is an item in the chunk buffer>)
```

Reply:
## Concern 3
What if there are a bunch of coroutines scheduled to run and deliver data to s3, but haven't run yet. This data will be lost.

Is there a way to identify the number of chunks on a plugin? I see we have input plugins

Maybe you could wait for the buffer space to drop to 0 indicating that all the data has been successfully enqueued into the s3 buffer. Then also check that the buffer is empty
fs_backlog_chunks_size: https://github.com/fluent/fluent-bit/blob/703da7d3ea78295fbed0560c25a8772b99cdc264/include/fluent-bit/flb_output.h#L366

So we have 3 places where logs can exist
1 Input - This seems to stop being ingested here: https://github.com/fluent/fluent-bit/blob/52fe1bdbf1c89daffe9a0a2542ae3696ebbdf849/src/flb_engine.c#L474-L477
2. As a Flush Coro that hasn't started yet or has started. - Most likely this is indicated by the fs_backlog_chunks_size field. 0, means that this data has been sent to (3.)
3. The S3 Buffer - Once processed, the data is sent to the s3 buffer. The buffer's linked list being empty would indicate that this data has been sent or dropped.

Let's make sure all data from 2 and 3 are processed (empty) before stopping the daemon coroutine, as it is needed to send out the last bits of data.

Reply:
So the revised while condition would be:
```
while (config->is_running == FLB_TRUE || x->fs_backlog_chunks_size != 0 || <there is an item in the chunk buffer>)
```

Reply:
I want to preserve the existing behavior which is that S3 never blocks the engine shutdown. Since S3 always buffers locally, any data it gets sent can be recovered.  Adding some sort of wait would introduce new shutdown blocking behavior.

Reply:
> If retries can still be triggered, however, then we may have an edge case where a new chunk could be ingested.

Its not. 

Once all flush coros are done, the engine will invoke cb_s3_exit which will try once to send all pending chunks. So the data still has a chance to be sent.

Reply:
Got it, if retries aren't processed, and we don't want the engine to be blocked on shutdown then terminating the daemon coroutine seems like a good idea.

Not sure if cb_s3_exit can use the async network stack.
Also it looks like flb_output_exit which invokes the cb_exit callback is called before the remaining coroutines finish. I suppose this is acceptable, because after the grace timeout, all logs that should be processed already had their chance to do so.
https://github.com/fluent/fluent-bit/blob/52fe1bdbf1c89daffe9a0a2542ae3696ebbdf849/src/flb_engine.c#L1042

Reply:
> Not sure if cb_s3_exit can use the async network stack.

I think its fine if it uses sync?

> Also it looks like flb_output_exit which invokes the cb_exit callback is called before the remaining coroutines finish. I suppose this is acceptable, because after the grace timeout, all logs that should be processed already had their chance to do so.

this is not what I see in the code... I see that its asking coros to stop happens first, but there's no wait/guarantee that they have finished before cb_exit...

though in my testing before when the daemon coro didn't end, I didn't see the cb_s3_exit being run... so I think we might be misreading the code here. 

If the exit callback and the daemon coro attempted to send data at the same time that could cause problems. 

I;'ll look more into this.

Reply:
This is a good point. Let me know what you find.

Reply:
These comments make sense to me. Shutting down when grace expires, even if some remaining coroutines are queued to add data to the buffer seems reasonable. We could say that at this point in time, after the grace is expired, that the remaining data should have already been sent out.

I think that the current way of doing things keeps things simple and is effective.